Issue #2974: Fix EmptyLineSeparator check when class members are separated by comments

jjlharrison opened the pull request:
Hi, 

I've attempted to fix issue #2974. I think I've managed to capture all of the edge cases by parsing comments from the beginning of the file. The code is quite complex, but I think that it needs to be due to all of the possible edge cases.

My test case covers all by 2 conditional branches in the new code. This means that the coverage check is failing.

I'm not expecting this pull-request to be merged at this stage, but as someone who hasn't contributed to Checkstyle before, I think I need some help with how to cover these 2 conditions. They only occur with invalid Java source files that have a single trailing slash at the end of a line. I can't work out how to add an invalid Java source file without the compiler complaining in the IDE. I tried giving the source file the extension ".txt", but the `TreeWalker` filters out files not ending with ".java".

As far as I am aware, all of the other checks are passing.

Thanks

jjlharrison commented:
FYI, the uncovered conditions are `lineHasNextChar()` on [EmptyLineSeparatorCheck:810](https://github.com/checkstyle/checkstyle/pull/5486/files#diff-f369e38e24af88245a4ff7ffea1cd55eR810) and [EmptyLineSeparatorCheck:826](https://github.com/checkstyle/checkstyle/pull/5486/files#diff-f369e38e24af88245a4ff7ffea1cd55eR826).

rnveach commented:
@jjlharrison Sorry for delay.

Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules. Squash all commits to 1 and keep them at 1.

Please fix TC error while I look into code coverage and shippable.
> EmptyLineSeparatorCheck.java (1)
890: getFlags() 'return' of Collection field flags

rnveach commented:
> uncovered conditions are lineHasNextChar() on EmptyLineSeparatorCheck:810

`lineHasNextChar()` never returns `false`. If we are always looking at a comment in this method, then the comment's start is always 2 characters `/*`.

> EmptyLineSeparatorCheck:826

`lineHasNextChar()` and `nextChar() == '/'` never returns `false`.
Comments can only be single line or multi-line. This method is called if the multi-line fails, so what else can this be but a single line comment?

To me it looks like you need to remove this extra code.

jjlharrison commented:
@rnveach:
> Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules. Squash all commits to 1 and keep them at 1.

I was aware of this rule, it was my intention to squash the commits and rebase once the problems have all been resolved.

> This method is called if the multi-line fails, so what else can this be but a single line comment?
> To me it looks like you need to remove this extra code.

That is correct, those conditions can only occur with invalid Java source code. I'm happy to remove the conditions, but I was trying to gracefully handle invalid files as I wasn't sure what the rules were for handling such input.

jjlharrison commented:
@rnveach: Ok, I think I've fixed the TC build, I've removed the redundant conditions, and I've squashed the commits. I'm hoping all of the checks will pass now.

rnveach commented:
@jjlharrison You plan to finish this? travis is showing an exception with regression. Please check it out and add a test to confirm it is fixed.
````
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 52
    at java.lang.String.charAt (String.java:658)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.nextChar (EmptyLineSeparatorCheck.java:874)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkStartComment (EmptyLineSeparatorCheck.java:811)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkStartStringLiteral (EmptyLineSeparatorCheck.java:802)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEndComment (EmptyLineSeparatorCheck.java:789)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEndStringLiteral (EmptyLineSeparatorCheck.java:773)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkQuoteEscape (EmptyLineSeparatorCheck.java:759)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEmpty (EmptyLineSeparatorCheck.java:745)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processChar (EmptyLineSeparatorCheck.java:735)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processLine (EmptyLineSeparatorCheck.java:711)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processLines (EmptyLineSeparatorCheck.java:697)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.categorizeLines (EmptyLineSeparatorCheck.java:622)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.hasEmptyLineBefore (EmptyLineSeparatorCheck.java:558)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.processPackage (EmptyLineSeparatorCheck.java:422)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.visitToken (EmptyLineSeparatorCheck.java:312)
````

jjlharrison commented:
@rnveach: I'm not sure what has happened to the Travis build, the commit it [built](https://travis-ci.org/checkstyle/checkstyle/jobs/333949905) is 354f7a0, my latest commit is 428feec. Is there any way to re-trigger the Travis build so it picks up the latest commit?

jjlharrison commented:
@rnveach: Awesome, found a way to test those conditions I was having problems testing with a valid file. I've change the test to reproduce the problem and fixed the bug 👌. Thanks.

rnveach commented:
@jjlharrison pitest is failing on whitespace package. Please generate report and fix the missing mutation it is asking for.
Let me know if you have any issues.

romani commented:
@jjlharrison , please finish PR, please rebase on latest master and we changed pitest validation a bit.

rnveach commented:
@jjlharrison whitespace package in pitest is still failing.
> `@@ -1 +1 @@`
> `-EmptyLineSeparatorCheck.java.html:<td class='covered'><pre><span class='survived'> if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {</span></pre></td></tr>`
> `+`
> `Actual:`
> `EmptyLineSeparatorCheck.java.html:<td class='covered'><pre><span  class='survived'>                if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {</span></pre></td></tr>`

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:
```diff
@@ -593,4 +610,276 @@ private static boolean isTypeField(DetailAST variableDef) {
+         */
+        private void checkStringLiteral() {
+            if (currentChar() == '"') {
+                if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {
```
this is the line pitest is complaining about

jjlharrison commented:
@romani @rnveach : Hope that does it. 🤞

rnveach commented:
@jjlharrison you can run the commands locally to verify it does fix the issue.
TC is now failing.
> 654: STRING_LITERAL_OPEN Field has no usages.

jjlharrison commented:
@rnveach: I am running the following locally to check:
```
man clean verify
mvn -e clean test org.pitest:pitest-maven:mutationCoverage -Ppitest-checks-whitespace
```

rnveach commented:
@romani CI is passing now.

romani commented:
@rnveach , please finish review of this PR, I will do review after you.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:
```diff
@@ -476,23 +478,25 @@ private boolean isViolatingEmptyLineBetweenFieldsPolicy(DetailAST detailAST) {
+                    if (previousLineFlags.contains(LineFlag.EMPTY)) {
+                        // 2 lines above the line being checked
+                        // and after the previous token are empty.
+                        result = true;
```
we should break out of the for loop as soon as we hit a result we are looking for.
there is nothing in this loop that will turn it back to `false`.

jjlharrison replied:
@rnveach : I think I was doing this in a previous revision, but when I do the build fails with the Checkstyle violation:
```
SingleBreakOrContinue: Loops should not contain more than a single "break" or "continue" statement
```

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:
```diff
@@ -551,9 +555,22 @@ private boolean hasEmptyLineBefore(DetailAST token) {
+                        // Line is empty.
+                        result = true;
+                    }
+                    if (lineFlags.contains(LineFlag.CODE)) {
```
shouldn't this be an `else if`?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:
```diff
@@ -593,4 +610,252 @@ private static boolean isTypeField(DetailAST variableDef) {
+    }
+
+    /**
+     * Parses a Java source file to assign each line a set of flags based on whether they have code,
```
Why are we parsing the Java source? It seems like you are attempting a crude attempt at understanding the source code when we have the ANTLR to do this for us.
Why can't we use the AST to understand the source?

jjlharrison replied:
I investigated that route, but it seemed to me that the AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines. If it is possible to change the way the AST is built to do this, the I agree, that would be better. Would you be able to point me in the right direction?

Also, prior to my changes the empty lines were detected by checking the lines with `CommonUtils.isBlank`, not by analysing the AST. I assumed this was because it couldn't be done with the AST.

rnveach replied:
> empty lines were detected by checking the lines with CommonUtils.isBlank, not by analysing the AST

Any empty line is an empty line. The util method is fine for this case. Emptyness could also be determined by looking for holes in the AST where nothing is on a specific line.

> AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines

This check is comment aware, so it knows about comments and javadocs. Comments aren't part of the `getAcceptableTokens` list, so `visitToken` won't stop on them by default but they are still in the tree.
Empty lines can be determined using `CommonUtils.isBlank`. The AST contains the line number.
Whitespace would have to be determined by looking at the line's column position and checking what comes before/after it.

> Would you be able to point me in the right direction?

I can but you would need to ask a more specific question.
Be aware, that comments are placed in weird locations sometimes, so it might be best that you add comments to the acceptable list to stop on them.

Example from issue:
````
    // Method 2 must fail
    public static final void method2() { }

    |--METHOD_DEF -> METHOD_DEF [8:4]
    |  |--MODIFIERS -> MODIFIERS [8:4]
    |  |  |--SINGLE_LINE_COMMENT -> // [7:4]
    |  |  |  `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
    |  |  |--LITERAL_PUBLIC -> public [8:4]
    |  |  |--LITERAL_STATIC -> static [8:11]
````

You can use http://checkstyle.sourceforge.net/cmdline.html#Command_line_usage option `-T` to print the AST with the comments to under the tree more.

jjlharrison replied:
@rnveach: Ok, I can look into this. I wish I had known about that before I had sunk hours into making all the CIs pass.

romani commented:
@rnveach , please finish your review.

@jjlharrison , please confirm that you addressed all review items and just wait for review.

rnveach commented:
@jjlharrison Please also make a comment when you have finished my items. We don't get notifications on new pushes to a PR.

rnveach commented:
@romani Review items are still not finished.

romani commented:
@jjlharrison , do you still want to finish this PR ?please finish.

romani commented:
@jjlharrison , please confirm that you want to stop this PR. It is completely ok, but it will let somebody else to continue your work.
If no reply till end of month, I will mark this PR as abandoned.

jjlharrison commented:
@romani: I intend to continue this when I have time, but I am quite busy at the moment, so if someone else wants to pick it up then they can go for it. Although, this issue has been open for over 2 years and had been inactive for a year when I started work on it, so it doesn't appear to high on anyones priority list.

romani commented:
@jjlharrison , thanks a lot for reply, at least I know that you are still willing to continue. I just try to keep open PRs under control, let me mark this PR as abandoned, if you will not find time till end of June, I will close PR, but code still be referenced in issue , you or anybody else can continue even you are always welcome. I just need to keep PR under control.

romani commented:
@jjlharrison , ping, please try to finish or at least share your problems.

jjlharrison commented:
@romani: The problems are already documented in this pull-request.

As I hinted at before, this pull request would have been finished by now if my time hadn’t been wasted by being asked to complete mutation coverage only to then be told fix this a different way.

romani commented:
>if my time hadn’t been wasted by being asked to complete mutation coverage

This is unfortunate, but there are case when all done and CIs are green,  but results of regression testing is not acceptable.

jjlharrison commented:
@romani: I accept that, and I agree with stringent quality checks, but it’s very demotivating and wasteful of time when this could have been discovered earlier rather than just blindly asking for CIs to be fixed without checking that what was being fixed was right.

Also, this isn’t a regression testing issue, the fix works, but there is just an implementation that is more consistent with how checks are normally written.

romani commented:
@jjlharrison , sorry that maintainers can not predict all problems at the beginning. We try to not read PRs that are failing because CI will lead contributor to better test code and in most cases it lead to conner cases that change solution completely. We can not read all versions of your implementation. But we do help when contributor ask questions and point to us that solution is not obvious and ....

Let's forget this unfortunate case. I will re-read your issue and code, we agree on solution first. Deal ?

romani commented:
@jjlharrison , please help me understand why you selected way with "parsing" of java code.

What I see is ... this Check is comments aware module, https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java#L255 , so it mean that it traverse Tree with comments. So all that you need is already in ASTree.

example from issue:
```
$ cat Test.java 
public class Test {

    public static final void method1() { }



    // Method 2 must fail
    public static final void method2() { }



    /**
     * Method 3 must fail
     */
    public static final void method3() { }
}

$ java -jar checkstyle-8.10.1-all.jar -T Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |   |--LITERAL_PUBLIC -> public [3:4]
    |   |   |--LITERAL_STATIC -> static [3:11]
    |   |   `--FINAL -> final [3:18]
    |   |--TYPE -> TYPE [3:24]
    |   |   `--LITERAL_VOID -> void [3:24]
    |   |--IDENT -> method1 [3:29]
    |   |--LPAREN -> ( [3:36]
    |   |--PARAMETERS -> PARAMETERS [3:37]
    |   |--RPAREN -> ) [3:37]
    |   `--SLIST -> { [3:39]
    |       `--RCURLY -> } [3:41]
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |   |--MODIFIERS -> MODIFIERS [8:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [7:4]
    |   |   |   `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
    |   |   |--LITERAL_PUBLIC -> public [8:4]
    |   |   |--LITERAL_STATIC -> static [8:11]
    |   |   `--FINAL -> final [8:18]
    |   |--TYPE -> TYPE [8:24]
    |   |   `--LITERAL_VOID -> void [8:24]
    |   |--IDENT -> method2 [8:29]
    |   |--LPAREN -> ( [8:36]
    |   |--PARAMETERS -> PARAMETERS [8:37]
    |   |--RPAREN -> ) [8:37]
    |   `--SLIST -> { [8:39]
    |       `--RCURLY -> } [8:41]
    |--METHOD_DEF -> METHOD_DEF [15:4]
    |   |--MODIFIERS -> MODIFIERS [15:4]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [12:4]
    |   |   |   |--COMMENT_CONTENT -> *\n     * Method 3 must fail\n      [12:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [14:5]
    |   |   |--LITERAL_PUBLIC -> public [15:4]
    |   |   |--LITERAL_STATIC -> static [15:11]
    |   |   `--FINAL -> final [15:18]
    |   |--TYPE -> TYPE [15:24]
    |   |   `--LITERAL_VOID -> void [15:24]
    |   |--IDENT -> method3 [15:29]
    |   |--LPAREN -> ( [15:36]
    |   |--PARAMETERS -> PARAMETERS [15:37]
    |   |--RPAREN -> ) [15:37]
    |   `--SLIST -> { [15:39]
    |       `--RCURLY -> } [15:41]
    `--RCURLY -> } [16:0]
```

attention to (there is clear gap between line  3 and 7-8, comments over method can be located a bit dipper in tree, but this is stable behavior):
```
    |       `--RCURLY -> } [3:41]
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |   |--MODIFIERS -> MODIFIERS [8:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [7:4]
    |   |   |   `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
```

Why we can not simply go over tree and find a gap that is more than 1 ? in example above gap is 3.

Ok, make example a bit complicated
```
$ cat Test.java 
public class Test {

    public void method1() { }

    // single line comment

    /** Some comment with empty line

    */
    // comment over method2
    public void method2() { }
}

$ java -jar checkstyle-8.10.1-all.jar -T Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |   `--LITERAL_PUBLIC -> public [3:4]
    |   |--TYPE -> TYPE [3:11]
    |   |   `--LITERAL_VOID -> void [3:11]
    |   |--IDENT -> method1 [3:16]
    |   |--LPAREN -> ( [3:23]
    |   |--PARAMETERS -> PARAMETERS [3:24]
    |   |--RPAREN -> ) [3:24]
    |   `--SLIST -> { [3:26]
    |       `--RCURLY -> } [3:28]
    |--METHOD_DEF -> METHOD_DEF [11:4]
    |   |--MODIFIERS -> MODIFIERS [11:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [5:4]
    |   |   |   `--COMMENT_CONTENT ->  single line comment\n [5:6]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [7:4]
    |   |   |   |--COMMENT_CONTENT -> * Some comment with empty line\n\n     [7:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [9:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [10:4]
    |   |   |   `--COMMENT_CONTENT ->  comment over method2\n [10:6]
    |   |   `--LITERAL_PUBLIC -> public [11:4]
    |   |--TYPE -> TYPE [11:11]
    |   |   `--LITERAL_VOID -> void [11:11]
    |   |--IDENT -> method2 [11:16]
    |   |--LPAREN -> ( [11:23]
    |   |--PARAMETERS -> PARAMETERS [11:24]
    |   |--RPAREN -> ) [11:24]
    |   `--SLIST -> { [11:26]
    |       `--RCURLY -> } [11:28]
    `--RCURLY -> } [12:0]
```

tracing a case:
```
    |--METHOD_DEF -> METHOD_DEF [3:4]
    ......
    |       `--RCURLY -> } [3:28]
    |--METHOD_DEF -> METHOD_DEF [11:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [5:4]
    |   |   |   `--COMMENT_CONTENT ->  single line comment\n [5:6]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [7:4]
    |   |   |   |--COMMENT_CONTENT -> * Some comment with empty line\n\n     [7:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [9:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [10:4]
    |   |   |   `--COMMENT_CONTENT ->  comment over method2\n [10:6]
```
previous method in on line 3
next method is on line 11
11 - 3 = 8 lines are some where 
look at comments
there is smth on line 5 ---- so line 4 is EMPTY !
theres is smth on line 7  ...... but nothing is on line 6 , so 6 is EMPTY !
block from line 7 finishes at line 9 , so line 8 is empty but is inside some block - skip
there is smth on line 10, ok
method is on line 11, ok
SO there is 2 EMPTY lines

looks simple , please let me know what I miss.

jjlharrison commented:
> So all that you need is already in ASTree.

@romani: Yes, I know the correct way to fix it, @rnveach [told me this back in March](https://github.com/checkstyle/checkstyle/pull/5486#discussion_r175460079).

> please help me understand why you selected way with "parsing" of java code

This fix I have submitted was mearly picking up where [others had left off](https://github.com/liscju/checkstyle/commit/d91467b132adda9c9fa216fb3b8c47074e4754e9) in issue #2974.

I completely agree that your suggested solution is the correct approach. I just don't have any time to dedicate to this at the moment. It's unfortunate, because I really would like this issue fixed.

romani commented:
too much mess, we need to drop this PR and start from scratch.

lets close  this PR, I already closed issue with reason https://github.com/checkstyle/checkstyle/issues/2974#issuecomment-401198949, to drop all connection to previous fixes.

@jjlharrison , you are welcome with better PR for newly opened issue.
